cmd/noms/diff/diff.go

```diff
@@ -5,121 +5,126 @@
+type Difference struct {
+	// Path to the Value that has changed
+	Path types.Path
+	// ChangeType indicates the type of diff: modified, added, deleted
+	ChangeType types.DiffChangeType
+	// OldValue is Value before the change, can be nil if Value was added
+	OldValue types.Value
+	// NewValue is Value after the change, can be nil if Value was removed
+	NewValue types.Value
+}
+
+// Differ is used to hold information necessary for diffing two graphs.
+type Differ struct {
+	// Channel used to send Difference objects back to caller
+	DiffChan chan<- Difference
+	// Channel that caller should close() to terminate Diff function.
+	StopChan chan struct{}
+	// Use LeftRight diff as opposed to TopDown
+	LeftRight bool
+}
+
+// Diff traverses two graphs simultaneously looking for differences. It returns
+// two channels: a DiffReceiveChan that the caller can use to iterate over the
+// diffs in the graph and a StopSendChanel that a caller can use to signal the
+// Diff function to stop processing.
+// Diff returns the Differences in depth-first first order. A 'diff' is defined
+// as one of the following conditions:
+//  * a Value is Added or Removed from a node in the graph
+//  * the type of a Value has changed in the graph
+//  * a primitive (i.e. Bool, Number, String, Ref or Blob) Value has changed
+//
+// A Difference is not returned when a non-primitive value has been modified. For
+// example, a struct field has been changed from one Value of type Employee to
+// another. Those modifications are accounted for by the Differences described
+// above at a lower point in the graph.
+//
+// If leftRight is true then the left-right diff is used for ordered sequences
+// - see Diff vs DiffLeftRight in Set and Map.
+//
+
+func (d Differ) Diff(v1, v2 types.Value) {
+	if !v1.Equals(v2) {
+		if !shouldDescend(v1, v2) {
+			d.sendDiff(Difference{Path: nil, ChangeType: types.DiffChangeModified, OldValue: v1, NewValue: v2})
+		} else {
+			d.diff(nil, v1, v2)
+		}
 	}
-
-	return diff(dw, types.Path{}, nil, v1, v2)
 }
 
-func diff(w diffWriter, p types.Path, key, v1, v2 types.Value) error {
+func (d Differ) diff(p types.Path, v1, v2 types.Value) bool {
 	switch v1.Type().Kind() {
 	case types.ListKind:
-		return diffLists(w, p, v1.(types.List), v2.(types.List))
+		return d.diffLists(p, v1.(types.List), v2.(types.List))
 	case types.MapKind:
-		return diffMaps(w, p, v1.(types.Map), v2.(types.Map))
+		return d.diffMaps(p, v1.(types.Map), v2.(types.Map))
 	case types.SetKind:
-		return diffSets(w, p, v1.(types.Set), v2.(types.Set))
+		return d.diffSets(p, v1.(types.Set), v2.(types.Set))
 	case types.StructKind:
-		return diffStructs(w, p, v1.(types.Struct), v2.(types.Struct))
+		return d.diffStructs(p, v1.(types.Struct), v2.(types.Struct))
 	default:
 		panic("Unrecognized type in diff function")
 	}
 }
 
-func diffLists(w diffWriter, p types.Path, v1, v2 types.List) (err error) {
+func (d Differ) diffLists(p types.Path, v1, v2 types.List) (cont bool) {
```

Comment:
I'm aware of avoiding more churn with this PR (so feel free not to do this) *but*:

Consistency wise, the other APIs we have tend to have a `stop` variable not a `cont` variable. E.g. Iter/IterAll, Walk. The advance is that you don't need to initialize `cont` to the common case - that you're *not* stopping.